crates/y-sweet/src/server.rs

```diff
@@ -101,38 +128,137 @@ impl Server {
+                    checkpoint_freq,
+                    doc_id,
+                    persistence_cancellation_token.clone(),
+                )
+                .instrument(span!(Level::INFO, "save_loop", doc_id=?doc_id_)),
+            );
 
-                    tracing::info!("Terminating loop.");
-                }
-                .instrument(span!(Level::INFO, "save_loop", doc_id=?doc_id)),
+            self.doc_worker_tracker.spawn(
+                Self::doc_gc_worker(
+                    self.docs.clone(),
+                    doc_id_.clone(),
+                    checkpoint_freq,
+                    persistence_cancellation_token,
+                    self.cancellation_token.clone(),
+                )
+                .instrument(span!(Level::INFO, "gc_loop", doc_id=?doc_id_)),
             );
         }
 
         self.docs.insert(doc_id.to_string(), dwskv);
         Ok(())
     }
 
+    async fn doc_gc_worker(
+        docs: Arc<DashMap<String, DocWithSyncKv>>,
+        doc_id: String,
+        checkpoint_freq: Duration,
+        persistence_cancellation_token: CancellationToken,
+        cancellation_token: CancellationToken,
+    ) {
+        let mut last_had_references = std::time::Instant::now();
+
+        loop {
+            tokio::select! {
+                _ = tokio::time::sleep(checkpoint_freq) => {
+                    if let Some(doc) = docs.get(&doc_id) {
+                        let awareness = Arc::downgrade(&doc.awareness());
+                        if awareness.strong_count() > 1 {
+                            last_had_references = std::time::Instant::now();
+                            tracing::debug!("doc is still alive - it has {} references", awareness.strong_count());
+                        } else {
+                            tracing::info!("doc has only one reference, candidate for GC");
+                        }
+                    } else {
+                        break;
+                    }
+
+                    if std::time::Instant::now() - last_had_references > checkpoint_freq * 2 {
+                        tracing::info!("GCing doc");
+                        docs.remove(&doc_id);
+                        break;
+                    }
+                }
+                _ = cancellation_token.cancelled() => {
+                    break;
+                }
+            };
+        }
+        persistence_cancellation_token.cancel();
+        tracing::info!("Exiting gc_loop");
+    }
+
+    async fn doc_persistence_worker(
+        mut recv: Receiver<()>,
+        sync_kv: Arc<SyncKv>,
+        checkpoint_freq: Duration,
+        doc_id: String,
+        cancellation_token: CancellationToken,
+    ) {
+        let mut last_save = std::time::Instant::now();
+
+        // There are two ways to trigger a save: either the dirty
+        // callback is called, or the cancellation token has been
+        // cancelled.
+        let mut is_done = false;
+        let mut interrupted = false;
```

Comment:
I find the flag variables make it hard to convince myself that this is right. This is psuedocode, but I think it reflects the desired behavior with a more linear flow:

```rust

loop {
    // wait until dirty
    select! {
        _ = recv.recv() => {
            // drop through to sleep
        }
        _ = cancellation_token.cancelled() {
             // no need to save because not dirty
             break
        }
    }

    // we are dirty; wait until throttled
    select! {
        _ = sleep_until_save() {
             save();
        }
        _ =  cancellation_token.cancelled() {
             save();
             break
        }
    }
}

```

Reply:
I'll try to poke at this again. The challenge is that we ~can't write `sleep_until_save` without getting into tricky futures/borrows-across-futures issues -- we don't want to move `recv` out of the loop and into a helper func.

how do you feel about manual `poll` implementations here?